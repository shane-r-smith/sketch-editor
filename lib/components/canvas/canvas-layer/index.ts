export * from "./canvas-element";
export * from "./canvas-layer";
export * from "./canvas-layer.types";
