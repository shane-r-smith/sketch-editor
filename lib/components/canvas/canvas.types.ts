import type { StageProps } from "react-konva";
import type { SketchElement } from "../../domain/sketch-element";
import type { Layer } from "../../domain";
import type { Stage } from "konva/lib/Stage";
import type { Ref } from "react";

export interface CanvasProps extends StageProps {
  sketchElement?: SketchElement;
  layerIndex?: number;
  layers: Layer[];

  ref: Ref<Stage>;
}
