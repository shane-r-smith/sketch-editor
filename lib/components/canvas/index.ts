export * from "./canvas-layer";

export * from "./canvas";
export * from "./canvas.types";
