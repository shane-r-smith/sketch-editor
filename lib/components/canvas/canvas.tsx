import { Stage } from "react-konva";
import type { CanvasProps } from "./canvas.types";
import { CanvasLayer } from "./canvas-layer";

export function Canvas({
  layers,
  layerIndex,
  element,
  ref,
  ...rest
}: CanvasProps) {
  return (
    <>
      <Stage {...rest} ref={ref}>
        {layers.map((layer) => (
          <CanvasLayer
            key={`layer_${layer.id}`}
            layer={layer}
            element={layerIndex === layer.index ? element : undefined}
          />
        ))}
      </Stage>
    </>
  );
}
