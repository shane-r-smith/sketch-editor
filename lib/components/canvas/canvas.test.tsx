import { render } from "@testing-library/react";
import { Canvas } from "./canvas";
import type { CanvasProps } from "./canvas.types";
import { vi } from "vitest";
import type { CanvasLayerProps } from "./canvas-layer";
import type { StageProps } from "react-konva";

// Mock CanvasLayer to avoid rendering actual Konva components
vi.mock("./canvas-layer", () => ({
  CanvasLayer: ({ layer, drawingElement }: CanvasLayerProps) => (
    <div data-testid="canvas-layer" data-layer-id={layer?.id}>
      {drawingElement && <div data-testid="canvas-element" />}
    </div>
  ),
}));

vi.mock("react-konva", () => ({
  Stage: ({ children }: StageProps) => (
    <div data-testid="konva-stage">{children}</div>
  ),
}));

describe("<Canvas />", () => {
  const mockRef = { current: null };
  const defaultProps: CanvasProps = {
    layers: [
      {
        id: "1",
        name: "layer 1",
        opacity: 0,
        elements: [],
        visible: false,
        locked: false,
      },
      {
        id: "2",
        name: "layer 2",
        opacity: 0,
        elements: [],
        visible: false,
        locked: false,
      },
    ],
    currentLayerIndex: 1,
    element: <div data-testid="custom-element" />,
    ref: mockRef,
    width: 400,
    height: 300,
  };

  it("renders correctly and matches snapshot", () => {
    const { container } = render(<Canvas {...defaultProps} />);
    expect(container).toMatchSnapshot();
  });
});
