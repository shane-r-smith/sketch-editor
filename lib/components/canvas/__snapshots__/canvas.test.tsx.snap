// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Canvas /> > renders correctly and matches snapshot 1`] = `
<div>
  <div
    data-testid="konva-stage"
  >
    <div
      data-layer-id="1"
      data-testid="canvas-layer"
    />
    <div
      data-layer-id="2"
      data-testid="canvas-layer"
    >
      <div
        data-testid="canvas-element"
      />
    </div>
  </div>
</div>
`;
